k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupPresenter.java

```diff
@@ -216,17 +226,44 @@ public void onNegativeClickedInConfirmationDialog() {
 
 
     private void autoConfiguration() {
-        view.setMessage(R.string.account_setup_check_settings_retr_info_msg);
-
         findProvider(accountConfig.getEmail());
     }
 
     private void findProvider(final String email) {
-        new AsyncTask<Void, Void, ProviderInfo>() {
+        findProviderTask = new AsyncTask<Void, Integer, ProviderInfo>() {
             @Override
             protected ProviderInfo doInBackground(Void... params) {
-                String[] emailParts = EmailHelper.splitEmail(email);
-                final String domain = emailParts[1];
+                publishProgress(R.string.account_setup_check_settings_retr_info_msg);
+
+                incomingReady = false;
+                outgoingReady = false;
+
+                final String domain = EmailHelper.getDomainFromEmailAddress(email);
+                ProviderInfo providerInfo = autoconfigureDomain(domain);
+
+                if (providerInfo != null || (incomingReady && outgoingReady)) {
+                    return providerInfo;
+                }
+
+                incomingReady = false;
+                outgoingReady = false;
+                try {
+                    String mxDomain = DNSHelper.getMXDomain(domain);
+
+                    if (mxDomain == null) return providerInfo;
```

Comment:
why not use the domain itself if there is no MX record? also, isn't providerInfo likely null here?